File: web/modules/voc.js
/* Voice of Customer — feedbacks dos alunos (elogios, reclamações, sugestões).
   Três abas: Dashboard (KPIs + gráficos + críticos + sugestões), Feedbacks
   (tabela filtrável + edição/exclusão) e Novo feedback (form inline).
   Feedback crítico dispara alerta no Slack. */
import { html } from "htm/preact";
import { useState } from "preact/hooks";
import { store } from "../lib/store.js";
import { useCollection, useUsers, useSelecao } from "../lib/hooks.js";
import {
  Badge, Table, Modal, Tabs, Kpi, BarChart, Text, Area, Select, FilterSelect,
  Assignee, BulkBar, aplicarEscopo, podeVerTudo, toast, baixarCSV, DOM,
} from "../lib/ui.js";
import { alertarSlack } from "../lib/notify.js";

const nowTs = () =>new Date().toISOString().slice(0, 19).replace("T", " ");
const vazio = (u) => ({ aluno: "", categoria: DOM.voc_categoria[0], tipo: "reclamação", gravidade: "Média", descricao: "", area: DOM.area[0], acao: "", status: "Aberto", responsavel: u ? u.nome : "" });

/* agrupa por chave e devolve [{label, n}, ...] ordenado desc, cortado em `top` */
function agrupar(rows, key, top) {
  const m = {};
  rows.forEach((r) => { const k = r[key] || "—"; m[k] = (m[k] || 0) + 1; });
  const arr = Object.entries(m).map(([label, n]) => ({ label, n })).sort((a, b) =>b.n - a.n);
  return top ? arr.slice(0, top) : arr;
}

export function View({ user }) {
  const { rows } = useCollection("voc", { order: "id", desc: true });
  const users = useUsers();
  const [aba, setAba] = useState("dashboard");

  return html`
    <h1 class="h1">Feedbacks</h1>
    <p class="sub">Feedbacks dos alunos: elogios, reclamações e sugestões — críticos alertam no Slack.</p>

    ${Tabs({ value: aba, onInput: setAba, options: [
      ["dashboard", "Dashboard"], ["lista", "Feedbacks"], ["novo", "Novo feedback"]] })}

    ${aba === "dashboard" ? html`<${Dashboard} rows=${rows}/>` : ""}
    ${aba === "lista" ? html`<${Lista} user=${user} users=${users} rows=${rows}/>` : ""}
    ${aba === "novo" ? html`<${Novo} user=${user} users=${users} onSalvo=${() =>setAba("lista")}/>` : ""}`;
}

/* ============================== DASHBOARD ============================== */
function Dashboard({ rows }) {
  if (!rows.length) return html`<div class="empty">Sem feedbacks registrados ainda.</div>`;
  const recl = rows.filter((r) =>r.tipo === "reclamação");
  const crit = rows.filter((r) =>r.gravidade === "Crítica");
  const sug = rows.filter((r) =>r.tipo === "sugestão");
  const sugPorCat = agrupar(sug, "categoria");

  return html`
    <div class="grid c3">
      ${Kpi({ label: "Total de feedbacks", val: rows.length })}
      ${Kpi({ label: "Reclamações", val: recl.length })}
      ${Kpi({ label: "Feedbacks críticos", val: crit.length, tone: "alert" })}
    </div>

    <div class="grid c2">
      <div class="card">
        <div class="section-t">Top 5 motivos de reclamação (por categoria)</div>
        ${recl.length ? BarChart({ data: agrupar(recl, "categoria", 5) }) : html`<div class="empty">Sem reclamações.</div>`}
      </div>
      <div class="card">
        <div class="section-t">Reclamações por área responsável</div>
        ${recl.length ? BarChart({ data: agrupar(recl, "area") }) : html`<div class="empty">Sem reclamações.</div>`}
      </div>
    </div>

    <div class="card">
      <div class="section-t">Feedbacks críticos</div>
      ${crit.length ? crit.map((r) =>html`<div class="h-item">
        ${Badge(r.tipo)} <b>${r.categoria}</b> · ${r.area}: ${r.descricao} ${Badge(r.status)}
      </div>`) : html`<span style="color:#9ca3af">Nenhum feedback crítico.</span>`}
    </div>

    <div class="card">
      <div class="section-t">Sugestões recorrentes</div>
      ${sugPorCat.length ? sugPorCat.map((s) =>html`<div class="h-item"><b>${s.label}</b>: ${s.n} sugestão(ões)</div>`)
        : html`<span style="color:#9ca3af">Sem sugestões registradas.</span>`}
    </div>`;
}

/* ================================ LISTA ================================ */
function Lista({ user, users, rows }) {
  const verTudo = podeVerTudo(user.perfil);
  const [escopo, setEscopo] = useState("minhas");
  const [edit, setEdit] = useState(null);
  const [fCat, setFCat] = useState("");
  const [fTipo, setFTipo] = useState("");
  const [fStatus, setFStatus] = useState("");

  const base = aplicarEscopo(rows, user, verTudo ? escopo : "minhas");
  const filtradas = base.filter((r) =>
    (!fCat || r.categoria === fCat) &&
    (!fTipo || r.tipo === fTipo) &&
    (!fStatus || r.status === fStatus));

  const { sel, toggle, toggleAll, clear } = useSelecao();
  const bulkResolver = async () => {
    const ids = [...sel];
    if (!ids.length) return;
    if (!confirm(`Marcar ${ids.length} feedback(s) como resolvido(s)?`)) return;
    await store.bulkUpdate("voc", ids, { status: "Resolvido" });
    await store.logAction(user.email, "voc_resolvidos_massa", `${ids.length} feedback(s)`);
    clear(); toast(`${ids.length} feedback(s) resolvido(s).`);
  };
  const bulkExcluir = async () => {
    const ids = [...sel];
    if (!confirm(`Excluir ${ids.length} feedback(s)? Essa ação não pode ser desfeita.`)) return;
    await store.bulkRemove("voc", ids);
    await store.logAction(user.email, "voc_excluidos_massa", `${ids.length} feedback(s)`);
    clear(); toast(`${ids.length} feedback(s) excluído(s).`);
  };

  const cols = [
    { key: "id", label: "ID" },
    { key: "aluno", label: "Aluno" },
    { key: "categoria", label: "Categoria" },
    { key: "tipo", label: "Tipo", render: (r) =>Badge(r.tipo), csv: (r) =>r.tipo },
    { key: "gravidade", label: "Gravidade" },
    { key: "descricao", label: "Descrição" },
    { key: "area", label: "Área" },
    { key: "responsavel", label: "Responsável" },
    { key: "status", label: "Status", render: (r) =>Badge(r.status), csv: (r) =>r.status },
    { key: "acao", label: "Ação" },
  ];

  return html`
    ${verTudo ? html`<div style="margin-bottom:14px">${Tabs({ value: escopo, onInput: setEscopo,
      options: [["minhas", "Minhas"], ["todas", "Equipe (todas)"]] })}</div>` : ""}

    <div class="toolbar">
      ${FilterSelect({ label: "Categoria", value: fCat, onInput: setFCat, options: DOM.voc_categoria })}
      ${FilterSelect({ label: "Tipo", value: fTipo, onInput: setFTipo, options: DOM.voc_tipo })}
      ${FilterSelect({ label: "Status", value: fStatus, onInput: setFStatus, options: DOM.voc_status })}
      <div class="grow"></div>
      <button class="btn" onClick=${() =>baixarCSV("voice_of_customer", cols, filtradas)}>Exportar</button>
    </div>
    <div class="count">${filtradas.length} feedback(s)${verTudo && escopo === "todas" ? " · equipe" : " · minhas"}</div>
    ${BulkBar({ n: sel.size, onClear: clear, actions: [
      { label: `Resolver (${sel.size})`, kind: "ok", on: bulkResolver },
      { label: `Excluir (${sel.size})`, kind: "danger", on: bulkExcluir },
    ] })}
    ${Table({ columns: cols, rows: filtradas, onRow: (r) =>setEdit({ ...r }), sel, onSel: toggle, onSelAll: toggleAll })}

    ${edit ? html`<${EditModal} user=${user} users=${users} fb=${edit} onClose=${() =>setEdit(null)}/>` : ""}`;
}

function EditModal({ user, users, fb, onClose }) {
  const [f, setF] = useState({ ...fb });

  const salvar = async () => {
    await store.update("voc", fb.id, { status: f.status, area: f.area, acao: f.acao, responsavel: f.responsavel });
    await store.logAction(user.email, "voc_editado", `#${fb.id}`);
    toast("Feedback atualizado."); onClose();
  };
  const excluir = async () => {
    if (!confirm("Excluir este feedback?")) return;
    await store.remove("voc", fb.id);
    await store.logAction(user.email, "voc_excluido", `#${fb.id}`);
    toast("Feedback excluído."); onClose();
  };

  return html`<${Modal} title=${"Feedback #" + fb.id} onClose=${onClose}
    footer=${html`<button class="btn danger" onClick=${excluir}>Excluir</button>
      <button class="btn" onClick=${onClose}>Fechar</button>
      <button class="btn primary" onClick=${salvar}>Salvar alterações</button>`}>
    <div class="h-item"><b>${fb.categoria}</b> / ${Badge(fb.tipo)} · Gravidade: ${fb.gravidade} · Aluno: ${fb.aluno || "N/I"}</div>
    <div class="h-item">${fb.descricao}</div>
    <div class="row c2">
      ${Select({ label: "Status", value: f.status, onInput: (v) =>setF({ ...f, status: v }), options: DOM.voc_status })}
      ${Select({ label: "Área responsável", value: f.area, onInput: (v) =>setF({ ...f, area: v }), options: DOM.area })}
    </div>
    ${Assignee({ value: f.responsavel, onInput: (v) =>setF({ ...f, responsavel: v }), users })}
    ${Text({ label: "Ação tomada", value: f.acao, onInput: (v) =>setF({ ...f, acao: v }) })}
  <//>`;
}

/* ================================= NOVO ================================= */
function Novo({ user, users, onSalvo }) {
  const [f, setF] = useState(vazio(user));

  const salvar = async () => {
    if (!f.descricao.trim()) return toast("Descreva o feedback.", "err");
    await store.insert("voc", { ...f, criado_em: nowTs() });
    await store.logAction(user.email, "voc_criado", `${f.categoria}/${f.tipo}`);
    if (f.gravidade === "Crítica") {
      alertarSlack(`:loudspeaker: *Feedback CRÍTICO* — ${f.categoria}/${f.tipo} (aluno ${f.aluno || "N/I"}, área ${f.area}): ${f.descricao.slice(0, 140)}`, "voc_critico");
      toast("Feedback crítico — alerta enviado ao Slack.", "warn");
    }
    toast("Feedback registrado.");
    setF(vazio(user));
    onSalvo();
  };

  return html`<div class="card">
    <div class="section-t">Registrar novo feedback</div>
    <div class="row c3">
      ${Text({ label: "Nome do aluno", value: f.aluno, onInput: (v) =>setF({ ...f, aluno: v }) })}
      ${Select({ label: "Categoria", value: f.categoria, onInput: (v) =>setF({ ...f, categoria: v }), options: DOM.voc_categoria })}
      ${Select({ label: "Tipo", value: f.tipo, onInput: (v) =>setF({ ...f, tipo: v }), options: DOM.voc_tipo })}
    </div>
    <div class="row c3">
      ${Select({ label: "Gravidade", value: f.gravidade, onInput: (v) =>setF({ ...f, gravidade: v }), options: DOM.gravidade })}
      ${Select({ label: "Área responsável", value: f.area, onInput: (v) =>setF({ ...f, area: v }), options: DOM.area })}
      ${Select({ label: "Status", value: f.status, onInput: (v) =>setF({ ...f, status: v }), options: DOM.voc_status })}
    </div>
    ${Assignee({ value: f.responsavel, onInput: (v) =>setF({ ...f, responsavel: v }), users })}
    ${Area({ label: "Descrição *", value: f.descricao, onInput: (v) =>setF({ ...f, descricao: v }) })}
    ${Text({ label: "Ação tomada", value: f.acao, onInput: (v) =>setF({ ...f, acao: v }) })}
    <div class="toolbar">
      <button class="btn primary" onClick=${salvar}>Registrar feedback</button>
    </div>
  </div>`;
}

File: modules/voc.py
"""Voice of Customer: registro e dashboard de feedbacks dos alunos."""
import pandas as pd
import streamlit as st

from core import db, ui
from core.slack import enviar_slack

CATEGORIAS = ["Plataforma", "Professor", "Horário", "Metodologia", "Comercial", "Financeiro", "Atendimento"]
TIPOS = ["elogio", "reclamação", "sugestão"]
GRAVIDADES = ["Baixa", "Média", "Alta", "Crítica"]
STATUS = ["Aberto", "Em tratativa", "Resolvido"]
AREAS = ["Produto", "Acadêmico", "Comercial", "Financeiro", "CX", "Diretoria"]


def render(user):
    ui.titulo("Voice of Customer", "Feedbacks dos alunos: elogios, reclamações e sugestões")
    tab_dash, tab_lista, tab_novo = st.tabs(["📊 Dashboard", "📋 Feedbacks", "➕ Novo feedback"])

    with tab_novo:
        with st.form("novo_voc", clear_on_submit=True):
            c1, c2, c3 = st.columns(3)
            aluno = c1.text_input("Nome do aluno")
            categoria = c2.selectbox("Categoria", CATEGORIAS)
            tipo = c3.selectbox("Tipo", TIPOS, index=1)
            c1, c2, c3 = st.columns(3)
            gravidade = c1.selectbox("Gravidade", GRAVIDADES, index=1)
            area = c2.selectbox("Área responsável", AREAS)
            status = c3.selectbox("Status", STATUS)
            descricao = st.text_area("Descrição *")
            acao = st.text_input("Ação tomada")
            if st.form_submit_button("Registrar feedback", type="primary"):
                if not descricao:
                    st.error("Descreva o feedback.")
                else:
                    db.execute(
                        "INSERT INTO voc (aluno,categoria,tipo,gravidade,descricao,area,status,acao,criado_em)"
                        " VALUES (?,?,?,?,?,?,?,?,?)",
                        (aluno, categoria, tipo, gravidade, descricao, area, status, acao, db.now()))
                    db.log_action(user["email"], "voc_criado", f"{categoria}/{tipo}")
                    if gravidade == "Crítica":
                        enviar_slack(f":loudspeaker: *Feedback CRÍTICO* — {categoria}/{tipo} "
                                     f"(aluno {aluno or 'N/I'}, área {area}): {descricao[:140]}", "voc_critico")
                        st.warning("Feedback crítico — alerta enviado ao Slack.")
                    st.success("Feedback registrado.")

    with tab_lista:
        c1, c2, c3 = st.columns(3)
        f_cat = c1.multiselect("Categoria", CATEGORIAS)
        f_tipo = c2.multiselect("Tipo", TIPOS)
        f_status = c3.multiselect("Status", STATUS)
        rows = db.query("SELECT * FROM voc ORDER BY id DESC")
        data = []
        for r in rows:
            if f_cat and r["categoria"] not in f_cat:
                continue
            if f_tipo and r["tipo"] not in f_tipo:
                continue
            if f_status and r["status"] not in f_status:
                continue
            data.append({
                "ID": r["id"], "Aluno": r["aluno"], "Categoria": r["categoria"],
                "Tipo": r["tipo"], "Gravidade": r["gravidade"], "Descrição": r["descricao"],
                "Área": r["area"], "Status": r["status"], "Ação": r["acao"],
            })
        df = pd.DataFrame(data)
        st.caption(f"{len(df)} feedback(s)")
        if not df.empty:
            st.dataframe(df, use_container_width=True, hide_index=True)
            ui.botoes_export(df, "voice_of_customer")

        st.divider()
        st.markdown("##### ✏️ Atualizar feedback")
        ids = [r["id"] for r in rows]
        if ids:
            sel = st.selectbox("Selecione", ids,
                               format_func=lambda i: next(f"#{r['id']} · {r['categoria']}/{r['tipo']}"
                                                          for r in rows if r["id"] == i))
            r = db.query_one("SELECT * FROM voc WHERE id=?", (sel,))
            with st.form("edit_voc"):
                c1, c2 = st.columns(2)
                status = c1.selectbox("Status", STATUS,
                                      index=STATUS.index(r["status"]) if r["status"] in STATUS else 0)
                area = c2.selectbox("Área responsável", AREAS,
                                    index=AREAS.index(r["area"]) if r["area"] in AREAS else 0)
                acao = st.text_input("Ação tomada", r["acao"] or "")
                if st.form_submit_button("Salvar", type="primary"):
                    db.execute("UPDATE voc SET status=?,area=?,acao=? WHERE id=?", (status, area, acao, sel))
                    db.log_action(user["email"], "voc_editado", f"#{sel}")
                    st.success("Atualizado.")
                    st.rerun()

    with tab_dash:
        rows = db.query("SELECT * FROM voc")
        if not rows:
            st.info("Sem feedbacks registrados ainda.")
            return
        df = pd.DataFrame([dict(r) for r in rows])
        recl = df[df["tipo"] == "reclamação"]

        c1, c2, c3 = st.columns(3)
        ui.card(c1, "Total de feedbacks", len(df))
        ui.card(c2, "Reclamações", len(recl), tipo="warn")
        crit = len(df[df["gravidade"] == "Crítica"])
        ui.card(c3, "Feedbacks críticos", crit, tipo="alert" if crit else "ok")

        st.write("")
        col_a, col_b = st.columns(2)
        with col_a:
            st.markdown("**Top 5 motivos de reclamação (por categoria)**")
            if not recl.empty:
                top = recl["categoria"].value_counts().head(5)
                st.bar_chart(top)
            else:
                st.caption("Sem reclamações.")
        with col_b:
            st.markdown("**Reclamações por área responsável**")
            if not recl.empty:
                st.bar_chart(recl["area"].value_counts())
            else:
                st.caption("Sem reclamações.")

        st.markdown("**🚨 Feedbacks críticos**")
        for r in db.query("SELECT * FROM voc WHERE gravidade='Crítica' ORDER BY id DESC"):
            st.markdown(f"- {ui.badge(r['tipo'])} **{r['categoria']}** ({r['area']}): "
                        f"{r['descricao']} — {ui.badge(r['status'])}", unsafe_allow_html=True)

        st.markdown("**💡 Sugestões recorrentes**")
        sug = df[df["tipo"] == "sugestão"]
        if not sug.empty:
            for cat, n in sug["categoria"].value_counts().items():
                st.markdown(f"- **{cat}**: {n} sugestão(ões)")
        else:
            st.caption("Sem sugestões registradas.")
